domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt

```diff
@@ -0,0 +1,74 @@
+package org.oppia.domain.audio
+
+import android.content.Context
+import android.media.MediaPlayer
+import android.net.Uri
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+class AudioPlayerController @Inject constructor(val context: Context) {
+
+  class PlayStatus(val type: String, val value: Int)
+
+  private var mediaPlayer: MediaPlayer? = null
+  private var executor: ScheduledExecutorService? = null
+  private var prepared = false
+  private val playState = MutableLiveData<PlayStatus>()
+
+  fun initializeMediaPlayer (stringUri: String) {
+    if (mediaPlayer == null) mediaPlayer = MediaPlayer()
+    mediaPlayer?.setOnCompletionListener {
+      stopUpdatingSeekBar()
+      playState.postValue(PlayStatus("COMPLETE", 0))
+    }
+    mediaPlayer?.setDataSource(context, Uri.parse(stringUri))
+    mediaPlayer?.prepareAsync()
+    mediaPlayer?.setOnPreparedListener {
+      prepared = true
+      playState.postValue(PlayStatus("DURATION", it.duration))
+    }
+  }
+
+  fun play() {
+    mediaPlayer?.let {
+      if (prepared && !it.isPlaying) it.start()
+      startUpdatingSeekBar()
+    }
+  }
+
+  fun pause() {
+    mediaPlayer?.let {
+      if (prepared && it.isPlaying) it.pause()
+    }
+  }
+
+  private fun startUpdatingSeekBar() {
+    if (executor == null) executor = Executors.newSingleThreadScheduledExecutor()
+    executor?.scheduleAtFixedRate({ updateSeekBar() }, 0, 1000, TimeUnit.MILLISECONDS)
```

Comment:
Who removes this schedule? Does MediaPlayer not provide a way for us to monitor the progress?

If not, we should use the background Kotlin dispatcher used by the DataProviders object. Unfortunately there isn't as nice of an API for scheduling at a fixed rate in Kotlin, but with ``delay()`` you can get close (and get some additional benefits I'll mention later):

```kotlin
suspend fun scheduleNextSeekBarUpdate() {
  nextUpdateJob = backgroundDispatcherScope.launch {
    delay(TimeUnit.SECONDS.toMillis(1)) // TODO: move delay to constant
    updateSeekBar()
    nextUpdateJob = scheduleNextSeekBarUpdate()
  }
  // NB: nextUpdateJob will need to be locked since it can be changed on multiple threads
}
```

Then, you can cancel the nextUpdateJob both during completion (like you similarly do with the executor service in stopUpdatingSeekBar below) and during deinitialization (which presumably is called by the UI when the user navigates away from the activity).

Reply:
From discussion & following up: I think we actually want a custom MutableLiveData class to schedule/stop in onActive/onInactive to ensure that we're only updating at times that observers actually care about it.